File: core/java/src/main/java/hara/lang/protocol/IStringLike.java
package hara.lang.protocol;

import hara.lang.declaration.HaraAvailability;
import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(
    namespace = "std.protocol.istringlike",
    name = "IStringLike",
    availability = HaraAvailability.CAPABILITY_GATED,
    capability = "native-runtime-protocols")
public interface IStringLike {
  @HaraMethod(value = "to-string", arity = 1)
  Object toStringValue();

  @HaraMethod(value = "from-string", arity = 2)
  Object fromString(Object value);
}

File: core/rust/src/lang/protocol/istringlike.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istringlike",
    name = "IStringLike",
    availability = "capability-gated",
    capability = "native-runtime-protocols"
)]
pub trait IStringLike {
    type Output;

    #[hara_method(value = "to-string", arity = 1)]
    fn to_string_value(&self) -> Self::Output;
    #[hara_method(value = "from-string", arity = 2)]
    fn from_string(&self, value: &str) -> Self::Output;
}
